### ISSUE TITLE
C++ Frontend data_parallel Does Not Update Weights

### ISSUE THREAD

**Issue by nmerrill67** [pytorch/pytorch#19540]
## 🐛 Bug

I have adapted the mnist.cpp example to work with two GPUs using torch::nn::parallel::data_parallel. However, the loss does not decay, and the accuracy never improves beyond 0.114 (the original single GPU example reaches around 0.99 accuracy).

## To Reproduce

1. Compile and run the attached example (change mnist_parallel.cpp.txt to mnist_parallel.cpp).
[mnist_parallel.cpp.txt](https://github.com/pytorch/pytorch/files/3100695/mnist_parallel.cpp.txt)
[CMakeLists.txt](https://github.com/pytorch/pytorch/files/3100696/CMakeLists.txt)

## Expected behavior

I would expect that the network could improve accuracy during training, but it does not.
## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
PyTorch version: 1.0.1
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 410.104
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.15.0
[conda] blas                      1.0                         mkl  
[conda] magma-cuda10              2.4.0                         1    cpbotha
[conda] magma-cuda100             2.5.0                         1    pytorch
[conda] magma-cuda90              2.5.0                         1    pytorch
[conda] mkl                       2019.3                      199  
[conda] mkl-include               2019.3                      199  
[conda] mkl-service               1.1.2            py36he904b0f_5  
[conda] mkl_fft                   1.0.10           py36ha843d7b_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] mkldnn                    0.16.1                        0    mingfeima
[conda] pytorch                   1.0.1           cuda100py36he554f03_0  
[conda] torchvision               0.2.1                    py36_0

## Additional context

If this is not a bug, and I am setting up the code incorrectly, please let me know and I will issue a feature request instead for a good data_parallel C++ example.

**Comment by nmerrill67** [comment]
Is there any update on this? This is a serious issue that has not been addressed at all. The DataParallel unit tests only test scatter and gather, but not with autograd. I believe that it does not work for training, which is displayed by my MNIST example that is attached above.

**Comment by soumith** [comment]
@yf225 @mrshenli can either of you take a look at it. Seems important enough to bump up priority

**Comment by mrshenli** [comment]
Let me investigate.

**Comment by mrshenli** [comment]
It looks like the [replicated module copies](https://github.com/pytorch/pytorch/blob/9fbce974c96a7768ce38f860bebcd9f17a66c9ec/torch/csrc/api/include/torch/nn/parallel/data_parallel.h#L41) are not [accumulating grads](https://github.com/pytorch/pytorch/blob/9fbce974c96a7768ce38f860bebcd9f17a66c9ec/torch/csrc/api/include/torch/nn/cloneable.h#L36) back to the original module (confirmed with @goldsborough ). I will implement `Broadcast` and `ReduceAddCoalesced` in C++ (as we did in Python). I feel eventually we might want to call C++ `data_parallel` from Python and drop Python's implementation.

**Comment by mrshenli** [comment]
@soumith @gchanan @ezyang 

I am trying to implement what `replicate.py` did in C++, but encountered a problem. The idea is to add `Broadcast` and `ReduceAdd` autograd functions to C++, broadcast every parameter from the original module to all destination devices, and then assign them to the module replicas. The assignment step seems require changes to every module class. Let's take `Linear` ([.h](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/linear.h), [.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/nn/modules/linear.cpp)) as an example. It keeps `weight` and `bias` `Tensor` objects (not reference) as member fields, which are assigned in `reset()` using the return value of `register_parameter`. In the replicate step, I can modify `Module`'s `parameter_` values and gradient edges to point to the broadcast source (by declaring replicate either as a friend function or class to `Module`). However, I cannot touch the `weight` and `bias` member fields in `Linear`, because replicate does not know it is a `Linear` module. As a result,  I cannot setup their gradient edges to point to the broadcast source. 

I had a discussion with @gchanan. To fix this problem, one way is to add a `Replicatable` friend class, but that would require me adding a new function to every module (e.g., `load_from_parameters_()`) to allow them to pick up the latest changes in `parameters_`, like what `Cloneable` did with `reset()`? This seems to be an overkill as we only need this in `data_parallel.h`. Any thoughts or suggestions?

**Comment by mrshenli** [comment]
Trying manually setting gradient functions.

**Comment by mrshenli** [comment]
Hi @nmerrill67, a fix is posted in #20910. If this is blocking you, please try pulling that PR and see if that works for you. That PR might not be able to land immediately, as we need to collect feedbacks to make sure that _hack_ is acceptable.

**Comment by nmerrill67** [comment]
@mrshenli Thanks for proposing a fix! I will be sure to try it ASAP. Also if anything comes to mind I will be happy to provide feedback to make the autograd broadcasting more streamlined.
